packages/editor/src/services/container.service.ts

```diff
@@ -0,0 +1,266 @@
+  return schema.properties !== undefined;
+};
+const isArray = (schema: JsonSchema): boolean => {
+  return schema.items !== undefined;
+};
+const deepCopy = <T>(object: T): T => {
+  return JSON.parse(JSON.stringify(object)) as T;
+};
+
+/**
+ * A Property wraps a JsonSchema and provides additional information
+ * like a label and the property key.
+ */
+export interface Property {
+  /**
+   * The label is a text donating a human readable name of the schema the property describes.
+   */
+  label: string;
+  /**
+   * The property is a text donating the schema key from which this property was created.
+   */
+  property: string;
+  /**
+   * The schema is the JsonSchema this property describes.
+   */
+  schema: JsonSchema;
+}
+
+/**
+ * Interface for describing result of an extracted schema ref
+ */
+interface SchemaRef {
+  uri: string;
+}
+
+/**
+ * Interface wraps SchemaRef
+ */
+interface SchemaRefs {
+  [id: string]: SchemaRef;
+}
+
+/**
+ * Returns the label of a container
+ * @param {Property} containerProperty
+ * @returns {string}
+ */
+export const getPropertyLabel =  (containerProperty: Property): string => {
+  return _.find(
+    [
+      containerProperty.schema.title,
+      containerProperty.label,
+      containerProperty.schema.id,
+      containerProperty.property
+    ],
+    n => !_.isEmpty(n)
+  );
+};
+
+/**
+ * Finding required definitions from parentSchema by using schema refs
+ *
+ * @param parentSchema The root schema
+ * @param allRefs All the refs of root schema
+ * @param schemaRefs The current schema refs
+ * @param extractedReferences Contains the definition attributes for the current schema
+ * @returns SchemaRefs all the required reference paths for subschema
+ */
+const findReferencePaths = (parentSchema: JsonSchema,
+                            allRefs: SchemaRefs,
+                            schemaRefs: SchemaRefs,
+                            extractedReferences: { [id: string]: string }
+): SchemaRefs => {
+  return _.reduce(
+    schemaRefs, (prev, schemaRefValue)  => {
+      let refs = _.pickBy(prev, _.flip(key => _.startsWith(key, schemaRefValue.uri)));
+      if (extractedReferences[schemaRefValue.uri]) {
+        refs = undefined;
+      }
+      if (!extractedReferences[schemaRefValue.uri]) {
+        extractedReferences[schemaRefValue.uri] = schemaRefValue.uri;
+        prev = _.omitBy(prev, value => value.uri === schemaRefValue.uri);
+      }
+      if (refs !== undefined) {
+        findReferencePaths(parentSchema, prev, refs, extractedReferences);
+      }
+      return prev;
+    },
+    allRefs
+  );
+};
+
+/**
+ * Calculate references that are used in parentSchema and add copy them into schema
+ *
+ * @param parentSchema root schema which is used to find all the schema refs
+ * @param schema current subschema without resolved references
+ * @returns JsonSchema current subschema with resolved references
+ */
+export const calculateSchemaReferences = (parentSchema: JsonSchema,
```

Comment:
This is the actual function that takes care of creating a self-contained schema, I'd call that way then, e.g. `makeSchemaSelfContained`